### ISSUE TITLE
Better AI war logic redux

### ISSUE THREAD

**Issue by RecursiveVision** [LoneGazebo/Community-Patch-DLL#7324]
I'm rewriting SelectBestApproachTowardsMajorCiv(), and in addition to significant performance fixes and better sanity checks to prevent stupid decisions, I'm also attempting to improve the AI's war selection logic.

I've had a couple ideas on how to improve this, and my own work has included rewriting loads of the terrible vanilla diplo AI logic on military strength evaluations.

Essentially my plan is to have the AI pick its approaches and generate the scores for each approach. Then if an AI's highest approach for a player is WAR, they'll sort by the player with the highest WAR approach first. 

However, I'd also like to add some military sanity checks to these approach scores; an additional proximity check is an obvious thought, but there are some other multipliers I could use help with.

A couple ideas I'd like some help from @ilteroi on I'll list here.

**Idea 1: Army Composition**
1. Evaluate whether a land attack, a sea attack, air or combined attack would be most effective, and;
2. Check whether the AI already has an available army powerful enough to carry out the attack (not already busy somewhere else);
3. If it does, increase the AI's desire to target this guy.

I'd need help figuring out how to do a check for 1 and 2. **It may also be helpful to evaluate comparative strengths, e.g. check our naval strength versus their naval strength.**

**Idea 2: City Defensibility**
On the forums there are some reported instances wherein the AI, despite facing an opponent with a poorly defended city with no walls, delays a long time before attacking. Some of this is on the diplomacy AI end; however, it'd be useful to have a function to evaluate whether the closest city(ies) to an AI are well-defended or not (terrain, buildings) and if a city can easily be captured, to prioritize that target.

There's also the flipside of this: an AI should be more cautious if its own cities are exposed to the opponent. 

**But in considering this it shouldn't get frozen in a loop between capturing an undefended city and being too wary about this same city being undefended, that's a challenging aspect of this.**

I'd need help figuring out how to evaluate city "exposure score" and make decisions based on this. I know there was a function for this at one point.

**Idea 3: Economic Value**
I know there already exists CvCity::updateEconomicValue() but Milae pointed out to me that it seems to place too much of an emphasis on resources compared to World Wonders and yields. There's also several "todo" notes in the function about needing to evaluate a resource's value to the AI and considering the potential value of empty plots (a resource might spawn there later). I think looking at how many of a city's plots have tile improvements on them would also be a good idea.

I wonder if you could take a look at this function and see what you think, @ilteroi?

The perfect target would essentially be someone the AI hates, has poor defenses, that they have a suitable army to attack, with high economic value.

**Comment by ilteroi** [comment]
i'm traveling, so this is from memory ...

1. army composition
* the military AI already checks which attack type (land/naval/combined) works best and keeps only the best N targets.
* AI also can do "sneak attacks". in that case the army is gathered in peacetime, and war is declared when the army is at the border.
* there is a function to check whether there is a good target to attack, and you claimed it's already being used?
* local strength is tracked through "dominance zones" around each city. but they are unreliable, because of limited unit visibility. in fact the AI cheats a little bit here to make up for its lack of intelligence.
* i don't think there is an easy way to check relative player military strength - how would you count an invisible army on the other side of the map?

2. defendability
* again the military AI considers how favorable the terrain is, city hitpoints etc.
* same code could be re-used to do a reverse check for "how exposed am i"
* anyway for the offensive side, if a city is a military target then it's good for attacking. no further checks needed (assuming that the defender is not too powerful)
* these reports of "wasted chances" are interesting but it may be that the AI needed some time to recruit their army, so city production logic could also be an issue - unit production priority really needs to be boosted in that situation.

3. city value
* economic factors are also considered for target selection
* but yes, probably the code could use some love, i'll take a look next week maybe
* on the other hand i don't think economy should play a big role in target selection

**Comment by RecursiveVision** [comment]
@ilteroi Thanks for the reply!

1. Army Composition
- Yes, you added HaveValidAttackTarget a while back to replace the busted cached attack target function that was blocking the AI from declaring war, and I've employed it elsewhere...but I was under the impression it being false meant there was no path to the target, based on the comments in the code. I guess those were obsolete, it's news to me that it also evaluates the defensability of the target. That explains a lot, actually. 

- But it also leads me to worry that it's causing the AI to be insufficiently aggressive, particularly in the early game, or overestimating the opponent's defensive capabilities. There are many other variables however, so I can't confirm it's the cause. I'll keep an eye on it as I rewrite all of this. Probably I'll remove it as a factor from SelectBestApproachTowardsMajorCiv() but add it to a separate "war selection/sanity" check.

- I looked and there are already functions that separately calculate military might by domain, this may be useful.

2. Defendability
- Where can I find this code?
- I rewrote all the code for military threat and target value etc. in the diplomacy AI, should be better.

3. City Value
- That'd be awesome, thanks. And it's not top on my list of considerations either, but all other things equal the AI should probably favor a city with 5 World Wonders over a city with 9 Iron on it when they already have a global monopoly. :)

**Comment by ilteroi** [comment]
take a look at EvaluateTargetApproach and UpdateAttackTargets for the larger picture

**Comment by RecursiveVision** [comment]
Thanks! I'll take a look.

**Comment by RecursiveVision** [comment]
@ilteroi Okay, I've looked through it!

So as I understand it the AI loops through all possible attack targets. The best attack target for each city is stored. If at peace, attack targets that are less than 1/3 of the score of the global best attack target are discarded.

I'd like to be able to differentiate between a "valid attack target" (the AI has any target whatsoever, regardless of how good, i.e. it can pathfind to the target) and a "good attack target" (what we currently have when at peace).

The idea being, if the AI has no valid target it will almost never declare war, whereas if it has no good target it will substantially decrease the war weight and prevent bribed wars/demands/etc. against that player.

There may be circumstances where diplo AI should declare war despite not having a good attack target, like when it wants to desperately stop someone from winning; currently it's setting weight to 0.

**Comment by RecursiveVision** [comment]
Additionally differentiating between "good" and "valid" attack targets allows diplomacy AI to decide whether it should stay at war with someone or try to make peace.

**Comment by ilteroi** [comment]
sounds reasonable.

however i would like to prevent futile attacks on "bad but valid" targets. declaring war may still be useful to plunder trade routes etc. what's your take on this?

**Comment by RecursiveVision** [comment]
@ilteroi My suggestion would be to add a "IsGood" flag or similar to the attack target structure. Then when looping through the targets, set the good flag to false if it's below the score threshold rather than discarding the target entirely. Do this regardless of war/peace status.

Military/tactical AI can then call HaveGoodAttackTarget() - or whatever you want to call it - and avoid futile attacks even when at war with a player, whereas diplomacy/deal AI can call both HaveValidAttackTarget() and HaveGoodAttackTarget() to determine whether war is possible, to adjust war weights, and determine peace willingness.

I have no objections to the tactical AI avoiding futile attacks - futile attacks are essentially awarding free XP to the enemy, after all.

One thing worth noting is that a target that's bad when attacking alone could become feasible if the AI is in a coop war with another nearby player, or has bribed someone to join in the fight, or simply has an ally nearby. I'm not sure if tactical AI knows the power of teamwork or not :)

**Comment by azum4roll** [comment]
Please don't forget the most important part:

"Can we reach our target?"

Applies to bribe evaluation too.

**Comment by RecursiveVision** [comment]
@azum4roll No worries. :)

**Comment by ilteroi** [comment]
i just tried your latest and immediately got a division by zero crash. not good.

CvDiplomacyAI.cpp line 41356: int iDurationPercent = (iTurnsPassed * 100 / iDuration);

which version is stable? please do not push untested stuff or create an experimental branch ...

**Comment by RecursiveVision** [comment]
@ilteroi https://github.com/LoneGazebo/Community-Patch-DLL/issues/7272

**Comment by RecursiveVision** [comment]
I don't have a stable version at the moment, rsi should be considered experimental, hence my warning. I will be debugging it soon now that I've learned how to do that!

**Comment by ilteroi** [comment]
well that is an inefficient way to publish that information, but understood.

also note that the parentheses don't do anything there, should probably be (iTurnsPassed*100)/iDuration

**Comment by RecursiveVision** [comment]
Well, I did explicitly title it "don't merge RSI yet" and tagged you in it. :)

I suppose I'll name the branch experimental or something for next time.

To be specific I only have three things left on my long checklist. Once everything is done I'll run several test games in the debugger and fix all crashes and other unintended behavior.

Problem is that my changes didn't really work well independently with how messy the code was; I rewrote large portions of it from scratch, and it necessitated this kind of larger overhaul where a lot of linked components are changed at once.

Yeah, I have a bunch of unnecessary parentheses I should remove...will look at that soon.

**Comment by ilteroi** [comment]
what i meant to say is, the parentheses may be required to avoid integer truncation or at least to make explicit what the compiler would do anyway. but it's important to have them in the right place :)

**Comment by RecursiveVision** [comment]
@ilteroi I see!

Is there any reason to have brackets in conditional statements, by the way? I've seen it in several places, e.g.

`if ((GetMeanness() > 8) || (GetDenounceWillingness() > 8) || (GetLoyalty() < 3))`

It made me wonder in paranoia if I should be adding brackets around all conditionals :)

Also, any response to my suggestion above?

**Comment by ilteroi** [comment]
which suggestion?

anyway there are clear rules on operator precedence (eg here: https://docs.microsoft.com/en-us/cpp/cpp/cpp-built-in-operators-precedence-and-associativity?view=msvc-160). (in)equality takes precedence over logical or. so the parentheses are not necessary in your example.

i'll take a look at the target selection now. your suggestion should work.

**Comment by ilteroi** [comment]
tangentially related

![image](https://user-images.githubusercontent.com/1358966/103249298-cf6a1c00-496e-11eb-997d-aefe1f08853f.png)

elizabeth is game :)

then again napoleon can also be quite generous

![image](https://user-images.githubusercontent.com/1358966/103249411-5ae3ad00-496f-11eb-846d-c39c0fd28685.png)

**Comment by RecursiveVision** [comment]
@ilteroi Just saw your latest push...thanks, this will be of use! :)

Fixed the crash and removed a bunch of unneeded parentheses on my end. I'm almost done with my changes, then the debugging begins. :)

So I noticed in your latest commit when you do multiplication and then division you enclose the multiplication part in parentheses...if multiplication and division have equal precedence left to right, is this necessary or is it just "to make explicit what the compiler would do anyway"?

For instance, would these two lines of code produce identical results?
```
vApproachScores[MAJOR_CIV_APPROACH_HOSTILE] += (vApproachBias[MAJOR_CIV_APPROACH_HOSTILE] * iTheirEra * GetVictoryCompetitiveness()) / 4;
```
```
vApproachScores[MAJOR_CIV_APPROACH_HOSTILE] += vApproachBias[MAJOR_CIV_APPROACH_HOSTILE] * iTheirEra * GetVictoryCompetitiveness() / 4;
```
